### ISSUE TITLE
Using props in class-based stores

### ISSUE THREAD

**Issue by FredyC** [mobxjs/mobx-react#744]
the issue is that (for the mechanical refactor to hooks) the class store needs access to an observable version of the props that autoupdates as new props arrive so it can have computed values (and/or actions) that depend both on the state and the props. AFAICT useLocalStore ensures that the props being received are a mutable mobx object being continuously updated, or am I mistaken?

_Originally posted by @spion-h4 in https://github.com/mobxjs/mobx-react/issues/722#issuecomment-517728120_

**Comment by FredyC** [comment]
I don't think it's actually possible to access props with a class store or at least I don't know-how. The problem is that the class store doesn't share closure with the component.

What you can do is utilize `useAsObservableSource` to create an observable variant of props and pass that to a constructed class.

```jsx
const Comp = ({ propA, propB }) => {
  const source = useAsObservableSource({ propA, propB })
  const [store] = React.useState(() => new ClazzStore(source))
}
```

```js
class ClazzStore {
  constructor(source) {
    this.source = source
  }
  @computed
  get something() {
    return `${this.source.propA} / ${this.source.propB}`
  }
}
```

Fairly verbose if you ask me, but I don't think there is some other "magical" way. The `useLocalStore` basically does the same, but it's slightly easier.

**Comment by spion-h4** [comment]
Because contrary to current popular belief, classes are a fine choice for this.

Maybe something like this?

```typescript
interface BaseStore<T> {
  public props:T;
}

function useClassStore<U extends BaseStore>(Klass: {new(t:T):U}, props: T) {
  let source = useAsObservableSource(props);
  let [state] = useState(() => new Klass(source))
  return state;
}
```

I'm assuming that the contents of the source reference returned on the first render will be updated every time useAsObservableSource is called? With hooks, you can never tell what the exact semantics are...

**Comment by FredyC** [comment]
Yea I suppose such hook will work. If you are asking about including such a hook to the library, I am don't think it's a good idea as it's very opinionated. At the moment you have a store that has a different constructor argument, you cannot use that. That can be important in case of dependencies on other stores.

> I'm assuming that the contents of the source reference returned on the first render will be updated every time useAsObservableSource is called? With hooks, you can never tell what the exact semantics are...

You are assuming right. With hooks, you can't, but with MobX it's very important to have a stable reference so it's safe to assume you get that.

> Because contrary to current popular belief, classes are a fine choice for this.

Would love to learn why. They are just more verbose in my eyes and require special handling due to decorators. I don't see any real benefit except it's defined "outside of the component". But you can as well have a custom hook that has `useLocalStore` inside for each store and outcome is pretty much the same.

**Comment by urugator** [comment]
You can also simply update whats neccessary every render:
```javascript
const [store] = React.useState(() => new Store())
store.setDep1(props.dep1);
store.setDep2(props.dep2);
```

**Comment by FredyC** [comment]
@urugator That sounds easier to you? :) We have MobX here and relying on imperative updates sounds fairly awkward. Besides, such operation is a side effect and as such should be in `useEffect`. That means it can get delayed. Suddenly, you have one value at props and outdated value in the class instance.

**Comment by urugator** [comment]
@FredyC It works exactly the same as `useAsObservableSource`:
```javascript
const [source] = React.useState(() => observable({}))
runInAction(() => {
   Object.assign(source, props);
});
```
It's impossible to get "reactive props" without imperative side effect (copying props to observables). We can do this because Mobx is synchronous, so once you call the setter (or `Object.assign()`) everything is up to date.
It is awkward, but no more than using `useAsObservableSource` or `useLocalStore`. Principle is really the same - copy props to a local observable and use it instead of props directly.
This is just a lot more transparent and easier to follow imo.

**Comment by FredyC** [comment]
> We can do this because Mobx is synchronous, so once you call the setter (or `Object.assign()`) everything is up to date.

Yea, but isn't also against "don't change state during a render"? Moreover that MobX is a synchronous and changing state in the middle of rendering sounds like something must go wrong. And yet it obviously works, at least now, but I wonder if that will still work in Concurrent mode 🤔 

> It is awkward, but no more than using `useAsObservableSource` or `useLocalStore`.

The awkwardness in my eyes comes from the fact that you must adapt your store (and do it for each one) to accept some outside values. In the case of `useLocalStore` you get the `source` as an argument and you build the store with it. It feels much more natural and straightforward. Sure you can derive the class and have it there, but that comes with some price as well.

**Comment by urugator** [comment]
> Yea, but isn't also against "don't change state during a render"

It's not kosher and hardly something to recommend.

> In the case of useLocalStore

I am not advocating use of classes for local component state (and neither `useLocalState`). 
In case of refactoring I would invest time into transforming local observable/computed/action into useState/useMemo/useCallback or moving it to (normalized) global app state (mobx class or whatever).
I don't see much of a point in refactoring to hooks, while making things more complicated (and actually still avoiding them).

**Comment by FredyC** [comment]
I am using MobX local store merely for its convenience as I don't need to bother with listing deps in effects and callbacks. That's a major selling point for me. I am not sure what "complications" you are mentioning. On the contrary, things are complicated with immutable state.

**Comment by urugator** [comment]
> don't need to bother with listing deps in effects and callbacks

You need to bother with listing deps to `useLocalStore`
How does local store help with effects?
Action depending on props is a bad idea - that should be another rule to follow.
Remains computed, which is actually rarely needed (in a sense that provided optimization is effective).

> complicated with immutable state

I have no issues with `useState(() => observable({}))` or `useState(() => new Store())`
But these can't be used if they depend on props - you either need "hack" or hooks, so I am suggesting hooks...

**Comment by spion-h4** [comment]
The only reason I'm considering this (mechanical) refactor is because the "new" React is the only one getting new features (for example, there is no ability to access multiple contexts from the old react, while useContext can be invoked multiple times). 

I'm absolutely fine with continuing to use classes otherwise; in addition, the new state management hooks provide significantly worse experience compared to MobX so I don't think they're an option.

For reference, the fully mechanical refactor is as follows:

Assuming the component MyComponent

* you extract the render to a function
* you do `let thisState = useLocalState(props => new MyComponentState(props), props);` at the top
* you replace `this` with `thisState` in the render method
* you rename the old `MyComponent` class into `MyComponentState`, remove `extends React.Component` from it and add a `constructor(public props:MyComponentProps) {}`

There is however the issue of double-rendering:

https://codesandbox.io/s/aged-snow-5u146

Try changing the props of the component; two renders will be triggered, one by react due to a change in props, another due to useObserver being triggered by that change

I'm not sure how to avoid that one.

**Comment by spion-h4** [comment]
@urugator the reason we use local observable and computed instead of local component state and "useMemo" is because they are an excellent fit. Typically components will receive a somewhat complex MobX model object via props; however they need better refined / transformed data (a view if you will) suitable for presentation. 

Using observable and computed ensure that changes in the model will result with minimal and efficient recomputation of the view data; if the view consists of multiple pieces, it can also ensure that the smallest amount of DOM nodes re-render and update.

**Comment by urugator** [comment]
> issue of double-rendering:

Yes this is disadvantage of hooks, there is no way to bailout (I wonder whether we will use generators one day, yielding the effects instead of calling hooks).

> useContext can be invoked multiple times

I used a HOC like: `consume({ store1: Context1, store2: Context2 })(({ store1, store2 }) => {})`
If the context value updates often (probably not with mobx, but it's common with immutable state), you may still want to put `memo` between `useContext` and actual component (or wrap render in `useMemo`).   

> useMemo and friends are absolutely inadequate and underpowered in comparison

Just keep in mind that reactive props is nothing but a state change inside computation:
```javascript
class PseudoComponent {
  @observable local; 
  @observable props;
  
  @computed a() {
    this.local + this.props.x;  
  }
  
  this.render = computed(props => {
    // changing state, breaking one-way data flow    
    runInAction(() => {
      this.props = props;
    })    
    return this.a;                       
  }, {
    equals: diffVdom,
  })
}
```
By moving `props` into constructor you will just move the problem one level up. Eventually you will realize that all deps must be provided upfront (which is kinda obvious considering computeds are argument-less), otherwise you need to memoize over arguments.
Memoization/diffing is perhaps slower (than binding), but allows you to have dynamic dependencies (deps unknown at the time of definition) and therefore dynamic component hierarchies.

**Comment by spion** [comment]
So far, hooks have been nothing but a disadvantage :) 

> Would love to learn why. They are just more verbose in my eyes and require special handling due to decorators. I don't see any real benefit except it's defined "outside of the component". But you can as well have a custom hook for each store that utilizes useLocalStore and outcome is pretty much the same except its bit cleaner.

In TypeScript classes are the only construct that at the same time has a type and a runtime representation. This gives them many advantages, one of which is that they can enable easy, type-safe dependency injection by representing both the interface and the implementation of a class. They can also just replace interfaces and act as type-safe DI token:

```typescript
class ActuallyAnInterface {
  field1!: string;
  field2!: stirng;
  methodReturningString!: () => string;
}
```

With emitDecoratorMetadata you can use this "classterface" as a type safe DI token.

The "easymode"  classes are especially neat since you don't have to immediately separate the interface from the implementation if there is no need for that yet.

In our "framework", through react context, we access the injector, which provides the application singletons as well as component level transients: https://github.com/hfour/h4bff/pull/57

You can read more about the framework here: https://hfour.github.io/h4bff/#/ although the building blocks and the docs on how they fit together are not fully complete yet.

**Comment by spion** [comment]
So I think I managed to avoid the double-updating issue by making the props mutable and turning off the staleness notification for the reaction temporarily while assigning the new props:

https://codesandbox.io/s/trying-to-use-hooks-794py

Not yet sure whether its a good idea to pass on the observed props directly - I think it might be better to keep react props object semantics but implement another hook called `useObservableProps(props)` that simply fetches the already created observable props in a `hObserver` style component.

**Comment by spion** [comment]
@urugator & @FredyC  I believe the above technique separates the "props receiving" phase from the "computing the VDOM" phase nicely. When the props change, mobx gets out of the way since the render is going to happen anyway and there is no need for reactions; in all other cases, mobx takes care of re-renders. If I had access to mobx internals I would describe this as "reaction is already scheduled to run" but without that bit disabling the event was my best idea.I believe this implementation mirrors the original mobx-react API and behavior more closely.

Do you think there is a place for `useObservableProps` in mobx-react(-lite)?

**Comment by FredyC** [comment]
@spion I am kinda failing to follow it here and what exactly are you proposing. How is it different from `useAsObservableSource`? Can we adapt that existing hook?

**Comment by spion** [comment]
Its different in that the component wrapper itself runs the hook to create an observable version of the props, and the useObservableProps only provides the mutable object that was already created by the wrapper. So you cannot use a different observable source.

It has to be this way because the reaction needs to be disabled during the props update, and the props update needs to happen outside of the tracking region of the reaction otherwise disabling doesn't work

I tried implementing useAsObservableSource that temporarily disables the reaction, however if that's done during the tracking stage it messes up the state of the reaction:

https://codesandbox.io/s/trying-to-use-hooks-n469q

Click on a+=1 to update the props of the inner component, then on counter += 1. The second click will not trigger a render of Add. If you do the two steps again, counter+=1 will trigger a re-render the second time. Not sure why this happens.

**Comment by FredyC** [comment]
Can you draft how would the `useObservableProps` looks like? I suppose the one from the sandbox, is not it? In case it is then I am sure we don't want anything depending on global variables with Concurrent mode coming.

```ts
let currentProps: any = null;

function useObservableProps<T>(_props?: T): T {
  if (currentProps != null) return currentProps;
  else throw new Error('useObservableProps not available for non-observer component');
}
```

And btw, the `observer` got changed in `mobx-react-lite@2` (published under `@next`) to account for Concurrent mode, so you might want to experiment with that first.

**Comment by spion** [comment]
Its not a micro-optimization if *every single prop update* causes two calls to the render function, is it?

**Comment by spion** [comment]
@FredyC  The global variable cannot be in an incorrect state, since its not possible for concurrent react to interrupt the section where its available or run it in parallel with other sections:

```typescript
currentProps = props; // this is where the global variable becomes available
    reaction.current.track(() => {
      try {
        rendering = fn(p);
      } catch (e) {
        exception = e;
      }
    });
    currentProps = null; // this is where its no longer available
```

Concurrent react can pause rendering of child components but it can't interrupt the toplevel function call of the current component, so the use of the global variable should be fine.

BTW thanks for pointing me to the updated code - I'll look into how it works and mess around with it :)

**Comment by FredyC** [comment]
> Its not a micro-optimization if _every single prop update_ causes two calls to the render function, is it?

You know, I am fairly convinced that with MobX it doesn't happen that often to have props changing like crazy. Usually, you would be passing some stable store there. Sure, couple props might be needed to change from time to time, but imo it's not that often as it might seem. And if a component is really that convoluted to have several sources of re-render, it might be worth to rethink that and split up.

> Concurrent react can only postpone the full rendering of child components, it cannot interrupt a standard function call.

Doesn't it also mean it can render several components at a time? I haven't really examined it that much, but "concurrent" sounds like "parallel" to me.

I think that automatically wrapping all props to observable might be overkill. It would be done no matter if you actually use those props. As such I don't think it should be the default behavior, but opt-in.

There is a hidden limitation. The `useAsObservableSource` needs a stable shape object. That means it would crash in case pass slightly different composition of props to an observed component. That's another argument why it shouldn't be the default behavior.

Also, I am kinda confused. In `useObservableProps` you are accepting `props`, but not doing anything with that. I guess some experimental leftover?

The name `useObservableProps` is slightly misleading as it doesn't use any hooks and as such can be called from anywhere (except it will fail with an error when done that).

Lastly, it feels somewhat opinionated. Personally, I am using `useAsObservableSource` not for props only, there are also values from context or even "calculated values" based on props. With your solution, I would gain some performance but lose that benefit. 

I am not saying it cannot become part of `mobx-react-lite`, but it certainly needs a good test coverage to figure out pitfalls.

**Comment by spion** [comment]
> Also, I am kinda confused. In useObservableProps you are accepting props, but not doing anything with that. I guess some experimental leftover?

Yeah that part is suboptimal :( The issue is you need to either pass the props or the type of the props - otherwise typescript wont know what type of object to return.

Ordering rules don't apply for this "non-hook" so its not quite a hook, agreed. I thought it would be less confusing if it was presented as a hook

My first version was cleaner because it simply presented a mutable, observable props object as the argument so there was no need of useObservableProps. Its quite clean, but potentially a huge change of behavior if the props object is passed as a whole anywhere.

I don't necessarily expect this to become a part of mobx-react or mobx-react-lite as it is - I wanted to bounce off a few ideas about observable props and class based stores and see if I can get help coming up with some API that approximates the class API but also makes sense in a hooks context. 

To be honest I am surprised to hear that people aren't using mobx state in class components - was everyone using setState instead? Seems like such a natural extension for UI state like e.g. filter toggles, dropdown expansion, currently selected items etc... is everyone keeping that state outside the components too?

**Comment by FredyC** [comment]
> To be honest I am surprised to hear that people aren't using mobx state in class components - was everyone using setState instead? Seems like such a natural extension for UI state like e.g. filter toggles, dropdown expansion, currently selected items etc... is everyone keeping that state outside the components too?

The setState isn't inherently bad for simple stuff really. The class components are a bigger burden to treat that state properly imo. Duplicating code, no easy way to share common stuff, forgetting to clean up and a bunch of others.

Personally, I still sometimes opt for `useState`, mostly in cases where observer just feels like overkill. In conjunction with [Immer](https://github.com/immerjs/immer), the `useState` becomes even easier for more robust stuff. The `useState` becomes annoying in a moment you need various setters to be passed around. Sure, there is a `useReducer` ... but no thanks :) With MobX this isn't a problem as the `useLocalStore` can carry actions included in a store. Or even class-based stores are fine in that matter.

I never liked decorators in classes and in the essence classes alone. It just did not feel right. So I was rather using `recompose` like crazy with 10+ compositions over the functional component. It was hell to debug that, but it certainly helped me to mentally shift to hooks faster because I wasn't used to classes anyway :)

**Comment by FredyC** [comment]
I think we have exhausted the topic here with loads of useful information. For further questions, please open a new issue.
